[Quirk] Allow popups from Office 365 to OneDrive after file uploads on iOS

vitorroriz opened the pull request:
#### 658ef7f5fecb61c441bfcc280a110874d6258309
<pre>
[Quirk] Allow popups from Office 365 to OneDrive after file uploads on iOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=302365">https://bugs.webkit.org/show_bug.cgi?id=302365</a>
<a href="https://rdar.apple.com/157794706">rdar://157794706</a>

Reviewed by Brent Fulgham.

On iOS/iPadOS, when users upload files to Office 365 (m365.cloud.microsoft),
the popup that shows the uploaded file on OneDrive is blocked.

When a user uploads a file, a valid user gesture is generated and the file
picker is presented. After the user selects a file, it is uploaded
asynchronously to Office 365. Once the upload completes, office.com attempts
to open a new tab displaying the uploaded file in the OneDrive web app
(onedrive.live.com). However, at this point the original user gesture has
expired, and no active user gesture exists to authorize opening the new tab.
When popup blocking is enabled in Settings (javaScriptCanOpenWindowsAutomatically
is false), the popup is blocked and the user never sees their uploaded file.

This adds an iOS/iPadOS-only quirk that allows popups from m365.cloud.microsoft
to onedrive.live.com. This quirk is only needed on iOS/iPadOS where Safari
doesn&apos;t have per-popup UI prompts like on macOS.

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::shouldABypassPopupBlockerForQuirk):
(WebCore::LocalDOMWindow::open):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsPopupFromMicrosoftOfficeToOneDrive const):
(WebCore::handleMicrosoftCloudQuirks):
(WebCore::Quirks::determineRelevantQuirks):
* Source/WebCore/page/Quirks.h:
(WebCore::Quirks::shouldAllowPopupFromMicrosoftOfficeToOneDrive const):
* Source/WebCore/page/QuirksData.h:

Canonical link: <a href="https://commits.webkit.org/303014@main">https://commits.webkit.org/303014@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/81c9291273fb33d947e9e003632b5433a69a7af0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/130900 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/3224 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/138/builds/41859 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/138328 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/82561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/132771 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/155/builds/3206 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/156/builds/3066 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/138328 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/82561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/133846 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/155/builds/3206 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/138/builds/41859 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/138328 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/155/builds/3206 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/138/builds/41859 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/81587 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/155/builds/3206 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/41859 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/140810 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/153/builds/2967 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/156/builds/3066 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/140810 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/3013 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/138/builds/41859 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/140810 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/138/builds/41859 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/56004 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/158/builds/3035 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/66430 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/2856 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/3056 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/2965 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1618,6 +1618,21 @@ bool Quirks::shouldUseEphemeralPartitionedStorageForDOMCookies(const URL& url) c
+{
+    if (!needsQuirks())
+        return false;
+
```
This is going to be kind of expensive, because every time a website passes through `LocalDOMWindow::allowPopUp()` it's going to do this host comparison. 

You should do the following:
1. Add another `bool` to `QuirksData.h` to quickly check if we need to do this work. Something like `shouldAllowPopupFromMicrosoftOfficeToOneDrive`. False by default.
2. Add a handler:

```
static void handleMicrosoftOfficeQuirks(QuirksData& quirksData, const URL& /* quirksURL */, const String& /* quirksDomainString */, const URL& documentURL)
{
    auto topDocumentHost = quirksURL.host();
    quirksData.shouldAllowPopupFromMicrosoftOfficeToOneDrive = documentURL.host().endsWithIgnoringASCIICase("m365.cloud.microsoft"_s);
}
```

3. Add a domain handler in `Quirks::determineRelevantQuirks`:

```
        // Not sure which of these is the right top level domain for this use case.
        { "microsoft"_s, &handleMicrosoftOfficeQuirks },
        { "office"_s, &handleMicrosoftOfficeQuirks },
```

Those first three steps make it so that the `LocalDOMWindow::allowPopUp()` only has to check a boolean as long as the web process is running instead of re-running the string comparison for every check for a popup on every website in the world.

4. Change your check below to this:

```
// You should only get here if Quirks::needsPopupFromMicrosoftOfficeToOneDrive was true earlier
return  targetURL.host().endsWithIgnoringASCIICase("onedrive.live.com"_s);
```

brentfulgham commented on Source/WebCore/page/Quirks.h:
```diff
@@ -240,6 +240,10 @@ class Quirks {
     WEBCORE_EXPORT bool shouldUseEphemeralPartitionedStorageForDOMCookies(const URL&) const;
 
+#if PLATFORM(IOS_FAMILY)
+    bool shouldAllowPopupFromMicrosoftOfficeToOneDrive(const URL& targetURL) const;
```
Also add:

`bool needsPopupFromMicrosoftOfficeToOneDrive() const`

brentfulgham commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1618,6 +1618,21 @@ bool Quirks::shouldUseEphemeralPartitionedStorageForDOMCookies(const URL& url) c
+    return false;
+}
+#endif
+
```
You should also add this to avoid doing more work than needed for the typical use case:

```
bool Quirks::needsPopupFromMicrosoftOfficeToOneDrive(const URL& targetURL) const
{
    if (!needsQuirks())
        return false;

    return m_quirksData.shouldAllowPopupFromMicrosoftOfficeToOneDrive;
}
```

brentfulgham commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2898,7 +2914,7 @@ ExceptionOr<RefPtr<WindowProxy>> LocalDOMWindow::open(LocalDOMWindow& activeWind
         return RefPtr<WindowProxy> { nullptr };
 
-    if (!firstWindow.allowPopUp()) {
+    if (!firstWindow.allowPopUp(*firstFrame, firstFrame->protectedDocument()->completeURL(urlString))) {
```
I think you're doing more work than you should here for the normal case where someone doesn't need this quirk.

I suggest something like:

```
URL targetURL = firstWindow.protectedDocument()->quirks().needsPopupFromMicrosoftOfficeToOneDrive() ? firstFrame->protectedDocument()->completeURL(urlString) : {};
if (!firstWindow.allowPopUp(*firstFrame, WTFMove(targetURL)) {
    ...
```

Note: `Document::completeURL` returns a new copy, so we shouldn't call it if we do not need to.

brentfulgham commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -405,6 +405,22 @@ bool LocalDOMWindow::allowPopUp(LocalFrame& firstFrame)
         || firstFrame.settings().javaScriptCanOpenWindowsAutomatically();
 }
 
+bool LocalDOMWindow::allowPopUp(LocalFrame& firstFrame, const URL& targetURL)
```
Should be `URL&& targetURL`

brentfulgham reviewed (changes requested):
This is very close, but I think it adds some perf issues that we want to avoid.

brentfulgham commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2865,6 +2865,17 @@ ExceptionOr<RefPtr<Frame>> LocalDOMWindow::createWindow(const String& urlString,
 }
 
+#if PLATFORM(IOS_FAMILY)
+static bool shouldABypassPopupBlockerForQuirk(const Document* document, const String& urlString)
```
Typo? `shouldABypassPopupBlockerForQuirks` -> `shouldBypassPopupBlockerForQuirk`.

brentfulgham reviewed (commented):
Looks great! Thank you for correcting this! r=me

brentfulgham reviewed (approved):
Thank you for taking the time to adjust those possible performance issues. r=me!

vitorroriz committed:
[Quirk] Allow popups from Office 365 to OneDrive after file uploads on iOS
https://bugs.webkit.org/show_bug.cgi?id=302365
rdar://157794706

Reviewed by Brent Fulgham.

On iOS/iPadOS, when users upload files to Office 365 (m365.cloud.microsoft),
the popup that shows the uploaded file on OneDrive is blocked.

When a user uploads a file, a valid user gesture is generated and the file
picker is presented. After the user selects a file, it is uploaded
asynchronously to Office 365. Once the upload completes, office.com attempts
to open a new tab displaying the uploaded file in the OneDrive web app
(onedrive.live.com). However, at this point the original user gesture has
expired, and no active user gesture exists to authorize opening the new tab.
When popup blocking is enabled in Settings (javaScriptCanOpenWindowsAutomatically
is false), the popup is blocked and the user never sees their uploaded file.

This adds an iOS/iPadOS-only quirk that allows popups from m365.cloud.microsoft
to onedrive.live.com. This quirk is only needed on iOS/iPadOS where Safari
doesn't have per-popup UI prompts like on macOS.

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::shouldABypassPopupBlockerForQuirk):
(WebCore::LocalDOMWindow::open):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsPopupFromMicrosoftOfficeToOneDrive const):
(WebCore::handleMicrosoftCloudQuirks):
(WebCore::Quirks::determineRelevantQuirks):
* Source/WebCore/page/Quirks.h:
(WebCore::Quirks::shouldAllowPopupFromMicrosoftOfficeToOneDrive const):
* Source/WebCore/page/QuirksData.h:

Canonical link: https://commits.webkit.org/303014@main
